src/utils/polygon.h

```diff
@@ -991,9 +992,40 @@ class Polygons
         }
     }
 
-    void fixSelfIntersections()
+    void resize(const coord_t& numerator, const coord_t divisor)
```

Comment:
Why does this function take a numerator and divisor, rather than just a `const Ratio scale` or something? Seems a bit confusing.

Reply:
I'll have to look into this, since I'm not the original author. I think @rburema wrote it, maybe he can comment on this.

Reply:
Maybe because of rounding errors. Why I tried your suggestion just now it didn't print any walls

Reply:
Tried out the following code changes. But I'm getting some strange results. Created a patch see code block below

```
Index: src/settings/types/Ratio.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/settings/types/Ratio.h	(revision 9420b87291d9817c29ccbbfe149a20ad6028a9c6)
+++ src/settings/types/Ratio.h	(date 1600289522015)
@@ -24,6 +24,10 @@
      */
     constexpr Ratio(double value) : value(value) {};
 
+    template <typename T>
+    constexpr Ratio(const T& numerator, const T& divisor)
+        : value(static_cast<double>(numerator / divisor)) {};
+
     /*
      * \brief Casts the Ratio instance to a double.
      */
@@ -85,4 +89,4 @@
 
 }
 
-#endif //RATIO_H
\ No newline at end of file
+#endif //RATIO_H
Index: src/utils/polygonUtils.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/utils/polygonUtils.cpp	(revision 9420b87291d9817c29ccbbfe149a20ad6028a9c6)
+++ src/utils/polygonUtils.cpp	(date 1600289060211)
@@ -1421,6 +1421,8 @@
     }
 
     const coord_t half_epsilon = (epsilon + 1) / 2;
+    const Ratio ratio(static_cast<coord_t>(1), epsilon);
+    const Ratio ratio_inv(epsilon, static_cast<coord_t>(1));
 
     // Shrink (making _near_ self-intersections into _actual_ self-intersecrtions), fix, grow back to original size.
     // Do this repeatedly with different offsets, so points that are close together do actually merge.
@@ -1428,9 +1430,10 @@
     for (const Point& translate_vec : translate_vecs)
     {
         thiss.translate(translate_vec);
-        thiss.scale(1, epsilon);
+
+        thiss.scale(ratio);
         ClipperLib::SimplifyPolygons(thiss.paths);
-        thiss.scale(epsilon, 1);
+        thiss.scale(ratio_inv);
         thiss.translate(-translate_vec);
     }
 
Index: src/utils/polygon.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/utils/polygon.h	(revision 9420b87291d9817c29ccbbfe149a20ad6028a9c6)
+++ src/utils/polygon.h	(date 1600289367278)
@@ -16,8 +16,10 @@
 
 #include <initializer_list>
 
-#include "IntPoint.h"
+#include "logoutput.h"
 #include "../settings/types/AngleDegrees.h" //For angles between vertices.
+#include "../settings/types/Ratio.h"
+#include "IntPoint.h"
 
 #define CHECK_POLY_ACCESS
 #ifdef CHECK_POLY_ACCESS
@@ -992,9 +994,9 @@
         }
     }
 
-    void scale(const coord_t& numerator, const coord_t divisor)
+    void scale(const Ratio& ratio)
     {
-        if (numerator == 1 && divisor == 1)
+        if (ratio == 1.)
         {
             return;
         }
@@ -1004,7 +1006,10 @@
         {
             for (Point& pt : thiss[p])
             {
-                pt = pt * numerator / divisor;
+                log("prev point x %d, y %d\n",pt.X, pt.Y);
+                log("Ratio: %d\n", ratio.value);
+                pt = pt * ratio;
+                log("point x %d, y %d\n",pt.X, pt.Y);
             }
         }
     }
Index: tests/settings/SettingsTest.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/settings/SettingsTest.cpp	(revision 9420b87291d9817c29ccbbfe149a20ad6028a9c6)
+++ tests/settings/SettingsTest.cpp	(date 1600288761207)
@@ -150,6 +150,14 @@
     EXPECT_DOUBLE_EQ(Ratio(0.01618), settings.get<Ratio>("test_setting")) << "With ratios, the input is interpreted in percentages.";
 }
 
+TEST_F(SettingsTest, CalculateRatio)
+{
+    const coord_t num = 10;
+    const coord_t div = 400;
+    Ratio ratio(num, div);
+    EXPECT_DOUBLE_EQ(0.025, ratio) << "With ratios, the input is interpreted in percentages.";
+}
+
 TEST_F(SettingsTest, AddSettingDuration)
 {
     settings.add("test_setting", "1234.5678");
```

Reply:
Fixed it with some additions to the Ratio with an addition of a numerator, divisor constructor and equality operators.

The ratio should be static_cast<double> before multiplication with the point